src/App.tsx

```diff
@@ -136,26 +189,50 @@ function App() {
     });
   }
 
+  const setText = useCallback((newText: string) => {
+    const model = editor?.getModel();
+    if (!model || !editor) return;
+    model.pushEditOperations(
+      editor.getSelections(),
+      [
+        {
+          range: model.getFullModelRange(),
+          text: newText,
+        }
+      ],
+      () => null
+    );
+    editor.setPosition({ column: 0, lineNumber: 0});
+  }, [editor]);
+
   function handleLoadSample() {
-    if (editor?.getModel()) {
-      const model = editor.getModel()!;
-      model.pushEditOperations(
-        editor.getSelections(),
-        [
-          {
-            range: model.getFullModelRange(),
-            text: rustpadRaw,
-          },
-        ],
-        () => null
-      );
-      editor.setPosition({ column: 0, lineNumber: 0 });
-      if (language !== "rust") {
-        handleChangeLanguage("rust");
-      }
+    setText(rustpadRaw);
+    if (language !== "rust") {
+      handleChangeLanguage("rust");
     }
   }
 
+  const handleUploadFile = useCallback(async (file: File) => {
+    const text = await file.text();
+    setText(text);
+    const newLanguage = getLanguage(file.name);
+    if (newLanguage !== language) handleChangeLanguage(newLanguage);
+  }, [language]);
+
+  const handleDownloadFile = useCallback(() => {
+    const model = editor?.getModel();
+    if (!model || !editor) return;
+    const text = model.getValue();
+    const languageExtension = getFileExtension(language);
+    const fileName = "unknown." + languageExtension;
+    const file = new File([text], fileName);
+    const url = URL.createObjectURL(file);
+    downloadUri(url, fileName);
+    URL.revokeObjectURL(url);
```

Comment:
I think you could encapsulate this part into the `download` function, which could just download a text string and file name. Also, for the filename, probably `rustpad.[ext]` would be better than `unknown.[ext]`?

Reply:
I would prefer to wrap `downloadUri` in a `downloadText` or similar if you want it out of there.

DownloadUri could be relevant for other topics, I would like to keep it as a seperate entity so people can look at it.

Creating a File based on text in JS, and a URI to it, is a fair bit more specific.

Reply:
oh and yeah, sure on rustpad.[ext].

I was wondering a bit at that point if we could use a potential upload filename, but quickly abandoned that, as it seems... problematic and errorprone to implement.